feat(ui): linkblock feature

Isaacprogi opened the pull request:
This PR is for issue #371 

# Demo URL
https://holdex-venture-studio-git-fork-isaacp-38e10e-holdex-accelerator.vercel.app/

Isaacprogi commented:
There are some part of the code that I commented out.(the part of fetching the OG data). If there is a better way fetch the OG data before mount that would be nice. At the moment, to get the OG data, you access the item data via prop and call the fetch function on mount to get the OG data which results to additional delays of using a spinner to indicate loading.

Addis-Belete commented:
@Isaacprogi Provide a URL to test the feature.

Isaacprogi commented:
> @Isaacprogi Provide a URL to test the feature.

What type of url exactly.

zolotokrylin commented:
@georgeciubotaru can you trigger deployment here please

markholdex commented:
@Isaacprogi Your build fails for the reason below. Please have a look. Thanks!

<img width="1311" alt="image" src="https://github.com/holdex/holdex-venture-studio/assets/46079535/ce0101e3-4c32-4945-9cd8-4e135aa9a993">

markholdex commented:
@Isaacprogi  Also, before we dive into the review, submit the amount of hours you spent resolving this Problem. Thanks!

Isaacprogi commented:
@markholdex , here's a breakdown of my work:

I incorporated SCSS where needed because I but ran into issues with some Tailwind properties not cooperating. Unsure why.
Started with  Pug template but switched to HTML syntax as the link image wasn't rendering in Pug.Still looking into that.
Spent 3 and a half hours on the component design. The remaining 2 and a half hours were dedicated to nailing down the right metadata as I had two options jsonlink or open graph to fetch the data. But can still use other options of setting up a node backend and use a node package.

Also the issue has been fixed, renamed linkBlock.svelte to LinkBlock.svelte.Deploy should work

markholdex reviewed (changes requested):
@Isaacprogi I was testing your implementation but can already notice multiple bugs. For example: 
1. Some of the links from tables or text context are also rendered as LinkBlock. It's not supposed to work like that. Only standalone links should be treated as LinkBlocks.
<img width="3350" alt="Screenshot 2024-01-03 at 15 57 16" src="https://github.com/holdex/holdex-venture-studio/assets/46079535/16b73371-825a-488c-8c7c-c53d337ed4b1">


2. Title and description of the LinkBlock is the same in every component. Which means it's hardcoded instead of it being pulled from the OG information of the link. 
<img width="1345" alt="Screenshot 2024-01-03 at 15 59 16" src="https://github.com/holdex/holdex-venture-studio/assets/46079535/bf9e7275-3ebe-4f6b-9988-048503ce97f3">

---

These are only a few I noticed right away. Please double-check and test your own results before asking for a review or claiming the solution is ready. We are not supposed to test your solution on bugs here. You're the one responsible for it to work and responsible for not pushing new bugs into existing code. Our job is to approve and merge your solution. Thanks!

Isaacprogi commented:
Okay,
Its clearer now
I understand
I didn't put the standalone and conceptual link into perspective.
Will get it working soonest.

Isaacprogi commented:
Issue has been resolved
![resolved](https://github.com/holdex/holdex-venture-studio/assets/42997350/8d248f16-9091-4879-a089-6c596f02a2f2)
just needed to verify the text
![function](https://github.com/holdex/holdex-venture-studio/assets/42997350/89059b5a-f0c4-4ca3-b12c-b61740eccf00)

working fine now

Isaacprogi commented:
Those without image will appear in this format
![without](https://github.com/holdex/holdex-venture-studio/assets/42997350/025ee29a-8146-41e8-9dd3-4809b39576eb)

georgeciubotaru commented on src/lib/components/BodyRenderer/inline.pug:
```diff
@@ -1,5 +1,6 @@
 Switch(condition="{item.type}")
   Case(value="link")
+   LinkBlock(item="{item}")
```
@Isaacprogi There is an explicit place to render `LinkBlock` elements., they are different from the `inline Links` elements.

Isaacprogi replied:
Okay, I will inspect

georgeciubotaru commented on src/routes/api/og-meta-data/+server.ts:
```diff
@@ -0,0 +1,20 @@
+
+export const GET: RequestHandler = async ({ fetch, url }) => {
+  try {
+    const apiKey = 'pk_81a92661d3647e25a0b14380e3381569467f8c68';
```
why sensitive data is placed in the codebase?

Isaacprogi replied:
Yes, you're right
I kept it there because pulled envs overrides the ones stored in codebase when ever the server is started.

Isaacprogi replied:
I will need access to store it from where they are pulled from

georgeciubotaru replied:
@Isaacprogi How about using an open API like `open-graph-scraper`? 
![image](https://github.com/holdex/holdex-venture-studio/assets/8217518/23385508-69dd-43f3-806b-8237aa6dc1be)

Isaacprogi replied:
Okay that will be better
Though about that initially
Thanks

georgeciubotaru reviewed (changes requested):
see comments

Isaacprogi commented:
I have made the changes
Made use of open-graph-scraper
I rendered the linkBlock in the link component instead. It takes the content of the link as a prop. It verifies if the text value associated with the link is just text or url. It displays as linkBlock if it is a url or returns the slot if its not.
![linkblock](https://github.com/holdex/holdex-venture-studio/assets/42997350/dacb7dcb-d626-414c-a562-f52ff4193ce6)

georgeciubotaru reviewed (changes requested):
see comment

Isaacprogi commented on src/lib/components/BodyRenderer/Blocks/link.svelte:
```diff
@@ -45,7 +46,8 @@
 
 {' '}
-<a
+<LinkBlock {item}>
```
Okay

Isaacprogi replied:
@georgeciubotaru this is where i got to render it
![rendered](https://github.com/holdex/holdex-venture-studio/assets/42997350/d1d7113b-e2bd-4cac-9708-1be36cb2e948)

markholdex reviewed (changes requested):
@Isaacprogi your implementation works which is a good sign. However, visually it doesn't look the way I expected. If I compare it with the design, I notice the proportions and paddings do not match. 

<img width="949" alt="image" src="https://github.com/holdex/holdex-venture-studio/assets/46079535/5bad56d4-b478-4614-9fa8-ed56455b7dc4">

<img width="1572" alt="image" src="https://github.com/holdex/holdex-venture-studio/assets/46079535/d2953365-ad0d-47be-84c9-f6d4ad2d09b9">

Isaacprogi commented:
Okay

Isaacprogi commented:
![proportion](https://github.com/holdex/holdex-venture-studio/assets/42997350/c5bf8d1b-4c49-4be2-a1d3-20299196c595)
This is what I have
Made the adjustments properly

georgeciubotaru commented:
Hey, @Isaacprogi This PR of yours already is up for 1 week. But should have been merged already and fixed within the first 4 hours.
You started developing this PR without having a full understanding of what you should do to accomplish the scope.
You have the last chance to prove your skills, by asking the right questions and resolving the right issue, otherwise I will be forced to reject this PR.

Thank you.

Isaacprogi commented:
You're right @georgeciubotaru. I didn't really get the scope initially.
Sorry about that.
Please aside the design been the way it ought to be, is there anything else?
Also for the the design above, it looks that way because I used exact dimensions from the figma
Which are not always accurate based on my experience.I just wanted to see the review of the work based on exact dimension of the figma.

Isaacprogi commented:
> Hey, @Isaacprogi This PR of yours already is up for 1 week. But should have been merged already and fixed within the first 4 hours. You started developing this PR without having a full understanding of what you should do to accomplish the scope. You have the last chance to prove your skills, by asking the right questions and resolving the right issue, otherwise I will be forced to reject this PR.
> 
> Thank you.

@georgeciubotaru sorry about that
That's a mistake from my side. I didn't get the scope initially.
Please aside the design been the way it ought to be, is there anything else?

Isaacprogi commented:
> @Isaacprogi You have to identify the place where to render this block and what elements should be considered as `linkBlock`.
> 
> As I mentioned you have `BodyRenderer` component that passes blocks to other components, just check the payload to see what blocks you need to parse, then find the right place to render it.

@georgeciubotaru when i console the blocks passed from the renderer to the item component, the links appear with the paragraph block of type paragraph. What elements should be considered as linkblock elements.

georgeciubotaru commented:
@Isaacprogi In the console logs should appear and your custom linkBlock, different than `paragraph`

Isaacprogi commented:
@georgeciubotaru I did a proper inspect on the the payload,
I discovered there are different kind of links

Links within paragraph
Links within table 
Embedded  google links
Other standalone links

The description mentioned "links not anchors" but i need to know what links to target in other to have the custom linkBlock.

zolotokrylin commented:
@Isaacprogi 
> I am tracing the issue all the way into the parser but before the implementation,
I need to know format of the link for linkblock?

you need to target links which are within paragraph, but when there are no other content surrounds that link
have a look at how Tally form is embedded or how Youtube video is embedded. You need to reproduce the same but for rich link UI component.

@georgeciubotaru correct me if I am wrong.

Isaacprogi commented:
That is what i have targeted from the start. The links in in the blocks are within paragraphs blocks and so to render them I have to render and do the implementation within the paragraph blocks as i did initially.
Except i will have to alter the parser function for the paragraph in order for the links for have its separate blocks outside paragraph.The section @markholdex  showed me is only found with the paragraph block data

markholdex commented:
@Isaacprogi I redeployed after your latest commits. Is it ready for review? I didn't see any requests to review after your commits. Did you QA properly on your side and you're confident it's all good? Let me know. Thanks!

Isaacprogi commented:
> @Isaacprogi I redeployed after your latest commits. Is it ready for review? I didn't see any requests to review after your commits. Did you QA properly on your side and you're confident it's all good? Let me know. Thanks!

@markholdex , With this last commit, everything is good. I'm confident. Done necessary testing as well. The only issue is that google drive links require some sort of authentication for their og data to be fetched but others work as as expected.

markholdex reviewed (changes requested):
@Isaacprogi any idea why the link highlighted from the bottom is not converted into a linkblock? 
<img width="969" alt="Screenshot 2024-01-30 at 16 57 26" src="https://github.com/holdex/holdex-venture-studio/assets/46079535/33a9d18b-4bcc-4918-afd1-fabe36213cbc">

Checking Google doc and it seems they both links and both separated from any other text. 

<img width="767" alt="Screenshot 2024-01-30 at 16 59 21" src="https://github.com/holdex/holdex-venture-studio/assets/46079535/1faee336-72ac-4def-9b60-1e189ca39229">

Isaacprogi commented:
> @Isaacprogi any idea why the link highlighted from the bottom is not converted into a linkblock? <img alt="Screenshot 2024-01-30 at 16 57 26" width="969" src="https://private-user-images.githubusercontent.com/46079535/300855423-33a9d18b-4bcc-4918-afd1-fabe36213cbc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY2Mjc0NzMsIm5iZiI6MTcwNjYyNzE3MywicGF0aCI6Ii80NjA3OTUzNS8zMDA4NTU0MjMtMzNhOWQxOGItNGJjYy00OTE4LWFmZDEtZmFiZTM2MjEzY2JjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTMwVDE1MDYxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJjOTNkOTQ1NjZlMjYyODAyZGIyZmMyOGNhOGY3ZDQ5NTU0OTdhOWU0OWM0NjM1YzQ3NzdiMzkwYjFkZTI3MGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.vtibg910rVBWBdcORzcG4OqhMy4S7nDEgRcnXKQ4rkw">
> 
> Checking Google doc and it seems they both links and both separated from any other text.
> 
> <img alt="Screenshot 2024-01-30 at 16 59 21" width="767" src="https://private-user-images.githubusercontent.com/46079535/300855991-1faee336-72ac-4def-9b60-1e189ca39229.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY2Mjc0NzMsIm5iZiI6MTcwNjYyNzE3MywicGF0aCI6Ii80NjA3OTUzNS8zMDA4NTU5OTEtMWZhZWUzMzYtNzJhYy00ZGVmLTliNjAtMWUxODljYTM5MjI5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTMwVDE1MDYxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZmN2E0MjlmZjg0MjNkMGRlY2YwZjY5OGEwMWU2MTg1Y2RjYjcwODlmZDJlNjk2M2I4MzZjMTJiZjcyNzYwZmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.U7cZ4oo1jDTRpvt-neromliKYrup-JWU9b1JdPAI5_g">

@markholdex The reason is because the one at the bottom is not part of the block(linktool).
Instead they exist in the paragraph block.When you check the paragraph block they are called links

markholdex reviewed (changes requested):
@Isaacprogi I updated the document and the link seems to render now. However, the image is now being displayed. 

<img width="957" alt="image" src="https://github.com/holdex/holdex-venture-studio/assets/46079535/1d888189-8778-4b2a-a709-5f4775f0b586">

Isaacprogi commented:
I countered that issue at some point but it worked after some refresh.I am
looking into that currently

On Tue, Jan 30, 2024, 4:36 PM Mark Curchin ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
>
> @Isaacprogi <https://github.com/Isaacprogi> I updated the document and
> the link seems to render now. However, the image is now being displayed.
> image.png (view on web)
> <https://github.com/holdex/holdex-venture-studio/assets/46079535/1d888189-8778-4b2a-a709-5f4775f0b586>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/holdex/holdex-venture-studio/pull/565#pullrequestreview-1851636198>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKIBMZW7MVCVJA75F2O6NSDYREHPBAVCNFSM6AAAAABBJ6S4CKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQNJRGYZTMMJZHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Isaacprogi commented:
> @Isaacprogi I updated the document and the link seems to render now. However, the image is now being displayed.
> 
> <img alt="image" width="957" src="https://private-user-images.githubusercontent.com/46079535/300869347-1d888189-8778-4b2a-a709-5f4775f0b586.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY2OTAyMDQsIm5iZiI6MTcwNjY4OTkwNCwicGF0aCI6Ii80NjA3OTUzNS8zMDA4NjkzNDctMWQ4ODgxODktODc3OC00YjJhLWE3MDktNWY0Nzc1ZjBiNTg2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTMxVDA4MzE0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI0YmMxNDAxZTAwMDZmZWJhNThjNmRmNWViNmM5ZmU0NTU3NjVjNGI2ZGRjNmY1ZTZhZDViOTM2MTg1NDEzODYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0Z6cYjS-IJvb67XKAh3GCuKP0lNHjNf8qiiTCl5kZNk">

@markholdex Please have a look at the format of the image url returned by the Og
![format](https://github.com/holdex/holdex-venture-studio/assets/42997350/02d8afa3-4d0b-4fde-b551-2450e506a1bd)
That's the reason that url the image appears broken

markholdex commented:
> > @Isaacprogi I updated the document and the link seems to render now. However, the image is now being displayed.
> > <img alt="image" width="957" src="https://private-user-images.githubusercontent.com/46079535/300869347-1d888189-8778-4b2a-a709-5f4775f0b586.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY2OTAyMDQsIm5iZiI6MTcwNjY4OTkwNCwicGF0aCI6Ii80NjA3OTUzNS8zMDA4NjkzNDctMWQ4ODgxODktODc3OC00YjJhLWE3MDktNWY0Nzc1ZjBiNTg2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTMxVDA4MzE0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI0YmMxNDAxZTAwMDZmZWJhNThjNmRmNWViNmM5ZmU0NTU3NjVjNGI2ZGRjNmY1ZTZhZDViOTM2MTg1NDEzODYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0Z6cYjS-IJvb67XKAh3GCuKP0lNHjNf8qiiTCl5kZNk">
> 
> @markholdex Please have a look at the format of the image url returned by the Og ![format](https://private-user-images.githubusercontent.com/42997350/301092607-02d8afa3-4d0b-4fde-b551-2450e506a1bd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcyMTQzOTgsIm5iZiI6MTcwNzIxNDA5OCwicGF0aCI6Ii80Mjk5NzM1MC8zMDEwOTI2MDctMDJkOGFmYTMtNGQwYi00ZmRlLWI1NTEtMjQ1MGU1MDZhMWJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA2VDEwMDgxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTczYzMwNmY1NDRhMWNkZDAwZDY5YjE2ZmMzMGZiOTE3MWM3OTk3MGE0Y2YxZjkyYTI1N2FlODIyNWMyZDRkZmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.NOp0ym6OvwV0qil1q7Ni5HVygpNKkHYT0bnNBYZHd60) That's the reason that the image appears broken. This part "https://holdex.io" at the beginning should not be there

You are right here. We will fix the problem on our side in a separate PR.

markholdex reviewed (changes requested):
@Isaacprogi please, double-check the proportion of the images that are inserted in the linkblock. The current one looks squished from sides. 
<img width="729" alt="Screenshot 2024-02-06 at 12 14 07" src="https://github.com/holdex/holdex-venture-studio/assets/46079535/480c8cab-a605-4b6c-af4d-ff5519dd8541">

You can compare it by looking here: https://www.opengraph.xyz/url/https%3A%2F%2Fholdex.io

Isaacprogi commented:
> @Isaacprogi please, double-check the proportion of the images that are inserted in the linkblock. The current one looks squished from sides. <img alt="Screenshot 2024-02-06 at 12 14 07" width="729" src="https://private-user-images.githubusercontent.com/46079535/302585442-480c8cab-a605-4b6c-af4d-ff5519dd8541.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcyMTUzNzksIm5iZiI6MTcwNzIxNTA3OSwicGF0aCI6Ii80NjA3OTUzNS8zMDI1ODU0NDItNDgwYzhjYWItYTYwNS00YjZjLWFmNGQtZmY1NTE5ZGQ4NTQxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA2VDEwMjQzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFlNmFjY2VkNmMyMGFmYTI5MzkzY2I4NzNjOGU3N2JiMGEyNDA2M2EyYmRjNWJjNjIyOTdlOWQyMDNhMDhkOWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.jRw7P-kWnttKbz8rmmlwHN5VR0rij3Z8GN-i_TJTGO4">
> 
> You can compare it by looking here: https://www.opengraph.xyz/url/https%3A%2F%2Fholdex.io

Okay

markholdex reviewed (changes requested):
@Isaacprogi missing an important aspect from the design and UX: the hover effect and the hover behavior. 

If you double-check designs, there is a hover state with its own style. A hover state means that the LinkBlock component should behave a specific outcome when hovering it and clicking on it. The best example is this link bookmarks from Notion.

https://github.com/holdex/holdex-venture-studio/assets/46079535/dec382cd-6c07-4cfe-8b38-808c424371f7

Isaacprogi commented:
> @Isaacprogi missing an important aspect from the design and UX: the hover effect and the hover behavior.
> 
> If you double-check designs, there is a hover state with its own style. A hover state means that the LinkBlock component should behave a specific outcome when hovering it and clicking on it. The best example is this link bookmarks from Notion.
> 
>  linkblock.mp4

Okay

markholdex reviewed (changes requested):
@Isaacprogi 
Hover states are better now but why you didn't decide to make the LinkBlock clickable? I know you asked me about it in DM but why didn't you check how Medium or Notion implements it?

---
Also, as I mentioned earlier, Discord DM it's not an appropriate place to ask these type of questions. The question relevant to the PR should be asked in the PR.

Isaacprogi commented:
> @Isaacprogi Hover states are better now but why you didn't decide to make the LinkBlock clickable? I know you asked me about it in DM but why didn't you check how Medium or Notion implements it?
> 
> Also, as I mentioned earlier, Discord DM it's not an appropriate place to ask these type of questions. The question relevant to the PR should be asked in the PR.

Isaacprogi commented:
> @Isaacprogi Hover states are better now but why you didn't decide to make the LinkBlock clickable? I know you asked me about it in DM but why didn't you check how Medium or Notion implements it?
> 
> Also, as I mentioned earlier, Discord DM it's not an appropriate place to ask these type of questions. The question relevant to the PR should be asked in the PR.

Oh, Sorry about that @markholdex 
Let me make it clickable

zolotokrylin commented:
@Isaacprogi are you done here? Please request a review if you are done.

georgeciubotaru commented on src/lib/components/BodyRenderer/Blocks/LinkBlock/linkblocklink.svelte:
@Isaacprogi This part is not DRY as you already have 99% of the functionality in the `Blocks/link.svelte`

georgeciubotaru commented on src/lib/components/BodyRenderer/Blocks/LinkBlock/linkblockskeleton.scss:
Also the component naming: You are already in the LinkBlock folder, so no need to repeat yourself

georgeciubotaru commented on src/lib/utils/isUrl.ts:
if not used, remove it

georgeciubotaru commented on src/routes/(pages)/c/mixins.pug:
```diff
@@ -50,7 +50,7 @@ mixin short-link(href, icon, title)
 mixin inline-link(href, title)
   a.short-link(
-    class="relative underline underline-offset-4 bg-accent1-default/15 text-accent1-default text-paragraph-l transition-colors hover:bg-accent1-default/25 focus:bg-accent1-default/25"
+    class="relative underline underline-offset-4 bg-accent1-default/15 text-accent1-default text-paragraph-l transition-colors hover:bg-rating-a focus:bg-accent1-default/25"
```
this change is out of scope

georgeciubotaru commented on src/lib/components/BodyRenderer/Blocks/LinkBlock/linkblock.scss:
same here, no Tailwind class it's used

Isaacprogi commented:
Okay

georgeciubotaru reviewed (changes requested):
see comments above

georgeciubotaru commented on src/lib/components/BodyRenderer/Blocks/LinkBlock/index.svelte:
```diff
@@ -0,0 +1,153 @@
+    }
+  }
+
+  let active: boolean = false;
+  let componentElement: HTMLElement;
+
+  onMount(() => {
+    fetchMetaTags(site);
+  });
+
+  const handleClickOutside = (event: MouseEvent) => {
+    if (componentElement && !componentElement.contains(event.target as Node)) {
+      active = false;
+    }
+  };
+
+  onMount(() => {
+    document.addEventListener('click', handleClickOutside);
+  });
+
+  const handleClick = () => {
+    active = !active;
+  };
+
+  function handleKeydown(event: KeyboardEvent) {
+    if (event.key === 'Enter') {
+      handleClick();
+    }
+  }
```
no need for this.

The active state is not persistent after refresh and also the block cannot have active state, so get rid of these and cleanup the code

georgeciubotaru commented on src/lib/components/BodyRenderer/Blocks/LinkBlock/index.svelte:
```diff
@@ -0,0 +1,153 @@
+  import { getContext } from 'svelte';
+  import { regExp } from '$components/BodyParser/utils';
+
+  let themeContext: any = getContext('theme');
```
move if together with the other declarations

georgeciubotaru commented on src/lib/components/BodyRenderer/Blocks/LinkBlock/linkblock.scss:
```diff
@@ -0,0 +1,78 @@
+}
+
+.link-details.padding-with-image {
+  @apply px-[16px] py-[12px];
```
there are classes for that, no need for hardcoded values `px-4` or `py-3`

georgeciubotaru commented on src/lib/components/BodyRenderer/Blocks/LinkBlock/skeleton.scss:
```diff
@@ -0,0 +1,38 @@
+}
+
+.linkblock-skeleton .right-side {
+  @apply p-[2px_0_2px_2px] h-full flex items-center justify-center min-w-[5rem] max-w-max flex-none;
```
same here

georgeciubotaru commented on src/lib/components/BodyRenderer/Blocks/LinkBlock/linkblock.scss:
```diff
@@ -0,0 +1,78 @@
+}
+
+.light-shadow {
+  @apply shadow-[0_0_0_2px_rgba(46,131,229,0.2)] border border-[rgba(46,131,229,1)];
```
please replace with the dedicated TW classes:

https://github.com/holdex/holdex-venture-studio/blob/main/src/lib/styles/variables.scss

Isaacprogi replied:
I removed the class entirely because they were for the active states

georgeciubotaru reviewed (changes requested):
see my comment

Isaacprogi commented:
> see my comment

Hello @georgeciubotaru the site doesn't display anything again. I keep getting this page
![page](https://github.com/holdex/holdex-venture-studio/assets/42997350/8fd33414-811b-45d2-9863-ee35a7334579)

georgeciubotaru reviewed (changes requested):
see comment

Isaacprogi commented on src/lib/components/BodyRenderer/Blocks/LinkBlock/index.svelte:
```diff
@@ -0,0 +1,126 @@
+        <p class="ellipsis text-t1 title">{title}</p>
+        <p class="ellipsis text-t3 description">{description}</p>
+        <div class="flex items-center text-sm font-medium leading-5">
+          <Link item={passedLink} let:text>
+            <TextWrapper {text} />
+          </Link>
```
Hello @georgeciubotaru 
I will need to use a separate link component with Sam functionality as the actually link component or should I use navigate to on the main element of the linkbkock instead of using a

Isaacprogi replied:
I am trying to retain the Link component because of how it works with the links provided to it instead using ellipsis

georgeciubotaru commented:
Log: error from Vercel
![image](https://github.com/holdex/holdex-venture-studio/assets/8217518/b16d7aef-6eba-40b6-91ac-1f28485b4fa1)

Isaacprogi commented:
@markholdex 
This url requires authentication as it it returns 401 https://drive.google.com/file/d/195pLp7ER8atug2fKGhEV17TDVPjaCxhq/preview

markholdex commented:
Tried after the latest updates. Visual-wise it looks as intended.

Isaacprogi commented:
> Tried after the latest updates. Visual-wise it looks as intended.

Okay,
Thanks

markholdex commented:
> Spent 3 and a half hours on the component design. An additional 3+ were dedicated to nailing down the right metadata making it a total of 7. I had two options jsonlink or open graph to fetch the data. But I can still use other options of setting up a node backend and use a node package.
> 
> Also the issue has been fixed, renamed linkBlock.svelte to LinkBlock.svelte

@Isaacprogi do we have the possibility to handle the URLs that are private?

markholdex commented:
@Isaacprogi why not just display the information similarly to how we display links without OG info? We already have the URL, we don't need the cover image, and if we can get the website title be querying the root domain OG info, that's enough for private links.

Isaacprogi commented:
> @Isaacprogi why not just display the information similarly to how we display links without OG info? We already have the URL, we don't need the cover image, and if we can get the website title be querying the root domain OG info, that's enough for private links.

Okay let me work on that

Isaacprogi commented:
> why not just display the information similarly to how we display links without OG info? We already have the URL, we don't need the cover image, and if we can get the website title be querying the root domain OG info, that's enough for private links.
@markholdex 
Should it be just for private links or both.You said links without OG info. Do you mean how those without cover image are displayed

Isaacprogi commented:
> > why not just display the information similarly to how we display links without OG info? We already have the URL, we don't need the cover image, and if we can get the website title be querying the root domain OG info, that's enough for private links.
> > @markholdex
> > Should it be just for private links or both.You said links without OG info. Do you mean how those without cover image are displayed

@markholdex please check the last commit

zolotokrylin commented:
@Isaacprogi can you confirm it is your last review request? 
We can't spend more time on back and forth on reviews.

Isaacprogi commented:
@zolotokrylin Yeah
This is my last commit.
Thanks

Isaacprogi commented:
> @zolotokrylin Yeah This is my review request Thanks

zolotokrylin commented:
@angelicawill can you please review. Thanks!

angelicawill reviewed (commented):
Triggering deployment

angelicawill commented on src/lib/components/BodyRenderer/Blocks/LinkBlock/index.svelte:
```diff
@@ -0,0 +1,117 @@
+  href={url}
+  target={isHoldexLink ? '_self' : '_blank'}
+  rel="noreferrer"
+  class={`link-block bg-l1 flex dark:bg-l2  border border-solid border border-l4 shadow-accent1-default rounded-xl
```
there are 2 border classes here

angelicawill commented on src/lib/styles/theme.scss:
```diff
@@ -1,11 +1,14 @@
 @import './variables.scss';
 
+
```
only whitespace change here

angelicawill commented on src/routes/(pages)/about/+page.svelte:
```diff
@@ -11,7 +11,7 @@
   $: ({ data: storeData } = $store);
   $: message = Parser.parse(storeData?.message as Message);
-</script>
+  </script>
```
whitespace change only

angelicawill commented on src/routes/(pages)/c/[slug]/[messageSlug]/+page.ts:
```diff
@@ -7,5 +7,5 @@ export const load: PageLoad = async ({ data }) => {
     community: {} as Community,
   });
-  return { store };
+    return { store };
```
next time, it's better to not have these kind of changes (whitespace changes) as it is not relevant to this PR and issue

Isaacprogi replied:
Alright

angelicawill reviewed (changes requested):
2 observations (tested via local build)

-When I click the linkblocks that are private links, it redirects to /c/announcements/article-components instead of the link itself. you can check from this video

https://github.com/holdex/holdex-venture-studio/assets/78342026/501030c3-b4b2-4c90-864b-d21ae651d90f

- the link icon for linkblocks with empty image doesnt match the design. please check the correct size of the link icon in the Figma file.
![CleanShot 2024-03-12 at 13 33 39](https://github.com/holdex/holdex-venture-studio/assets/78342026/7eb62959-d43b-4ce1-9fd6-5d9d3604ae2c)

angelicawill commented on src/lib/components/BodyRenderer/inline.svelte:
```diff
@@ -12,7 +12,7 @@
   type Item = {
     type: string;
-    text: string;
+    text: string
```
why did you remove the semicolon?

Isaacprogi replied:
oh, that would probable be a mistake. Not sure when I did that

angelicawill commented on src/routes/(pages)/c/[slug]/[messageSlug]/+page.server.ts:
```diff
@@ -7,7 +7,6 @@ export const load: PageServerLoad = async ({ locals, params }) => {
     params.slug,
     params.messageSlug
   );
-
```
please watch for whitespace changes

angelicawill commented on src/lib/components/BodyRenderer/Blocks/LinkBlock/index.svelte:
```diff
@@ -0,0 +1,118 @@
+
+  $: title = (metaInfo?.meta?.title ?? '') as string;
+  $: description = (metaInfo?.meta?.description ?? '') as string;
+  $: url = (metaInfo?.link ?? item.url ?? "") as string;
```
please match the formatting of the remaining codebase, singlequote must be used for string instead of doublequote

angelicawill reviewed (changes requested):
Some minor code formatting issues, otherwise the functionality is working fine 👍.

Please carefully check for whitespace changes in your PR, even parts that were not mentioned by me, thanks!

Isaacprogi commented:
> Some minor code formatting issues, otherwise the functionality is working fine 👍.
> 
> Please carefully check for whitespace changes in your PR, even parts that were not mentioned by me, thanks!

Alright, thank you

angelicawill commented:
Don't forget to explicitly request for review if it's ready for review

Isaacprogi commented:
> Don't forget to explicitly request for review if it's ready for review

Okay

angelicawill commented:
@markholdex @georgeciubotaru can you help trigger preview deployment? thanks!

Isaacprogi commented:
Hello @zolotokrylin

zolotokrylin commented:
@markholdex @georgeciubotaru

markholdex commented:
Redeployed. Sorry, for some reason I was not getting notifications.

markholdex commented:
@Isaacprogi check this comment from the testimonial component that was implemented. https://github.com/holdex/holdex-venture-studio/pull/570#issuecomment-1900785030

Do you believe you know what is required to make the adjustments?

Isaacprogi commented:
> @Isaacprogi check this comment from the testimonial component that was implemented. [#570 (comment)](https://github.com/holdex/holdex-venture-studio/pull/570#issuecomment-1900785030)
> 
> Try to think about what is needed for your linkblock component. Do you believe you know what is required to make the adjustments?

You are  looking for a situation where all the logic of getting the data for the linkblock component is handled by the api/google-conversion if  I am right?

Isaacprogi commented:
> > @Isaacprogi check this comment from the testimonial component that was implemented. [#570 (comment)](https://github.com/holdex/holdex-venture-studio/pull/570#issuecomment-1900785030)
> > Try to think about what is needed for your linkblock component. Do you believe you know what is required to make the adjustments?
> 
> @markholdex You are looking for a situation where all the logic of getting the data for the linkblock component is handled by the api/google-conversion if I am right? So the data for the linkblock component should be ready in the blocks and imported in the linkblock instead of having all the logic of fetching the data for the linkblock compoenent in the linkblock component.

please more clarity

Isaacprogi commented:
> > > @Isaacprogi check this comment from the testimonial component that was implemented. [#570 (comment)](https://github.com/holdex/holdex-venture-studio/pull/570#issuecomment-1900785030)
> > > Try to think about what is needed for your linkblock component. Do you believe you know what is required to make the adjustments?
> > 
> > 
> > @markholdex You are looking for a situation where all the logic of getting the data for the linkblock component is handled by the api/google-conversion if I am right? So the data for the linkblock component should be ready in the blocks and imported in the linkblock instead of having all the logic of fetching the data for the linkblock compoenent in the linkblock component.
> 
> please more clarity

@georgeciubotaru

markholdex commented:
> > @Isaacprogi check this comment from the testimonial component that was implemented. [#570 (comment)](https://github.com/holdex/holdex-venture-studio/pull/570#issuecomment-1900785030)
> > Try to think about what is needed for your linkblock component. Do you believe you know what is required to make the adjustments?
> 
> @markholdex You are looking for a situation where all the logic of getting the data for the linkblock component is handled by the api/google-conversion if I am right? So the data for the linkblock component should be ready in the blocks and imported in the linkblock instead of having all the logic of fetching the data for the linkblock compoenent in the linkblock component.

@Isaacprogi I think you have the right overall idea. If you are unsure you can also look at many other examples available in the repo. Are you available to wrap it up?

Isaacprogi commented:
> > > @Isaacprogi check this comment from the testimonial component that was implemented. [#570 (comment)](https://github.com/holdex/holdex-venture-studio/pull/570#issuecomment-1900785030)
> > > Try to think about what is needed for your linkblock component. Do you believe you know what is required to make the adjustments?
> > 
> > 
> > @markholdex You are looking for a situation where all the logic of getting the data for the linkblock component is handled by the api/google-conversion if I am right? So the data for the linkblock component should be ready in the blocks and imported in the linkblock instead of having all the logic of fetching the data for the linkblock compoenent in the linkblock component.
> 
> @Isaacprogi I think you have the right overall idea. If you are unsure you can also look at many other examples available in the repo. Are you available to wrap it up?

@markholdex Okay, yeah i'm available to wrap it up

Isaacprogi commented:
@markholdex , please do i need to move the fetch  to api/google-conversion as well?

markholdex commented:
> @markholdex , please do i need to move the fetch to api/google-conversion as well?

@Isaacprogi please check with the examples available in the repository and the URL I sent you and take the decision yourself. Part of the trial is also to evaluate your decision-making and judging skills. You must take responsibility for what you are executing.

Isaacprogi commented:
> > @markholdex , please do i need to move the fetch to api/google-conversion as well?
> 
> @Isaacprogi please check with the examples available in the repository and the URL I sent you and take the decision yourself. Part of the trial is also to evaluate your decision-making and judging skills. You must take responsibility for what you are executing.

@markholdex Please provide link to more aligned and relevant examples.I have made the changes already.Just need to be sure that what I did is right.

markholdex commented:
@Isaacprogi you can double-check with relevant PR here: https://github.com/holdex/holdex-venture-studio/pull/570#issuecomment-1900785030 where we implemented a similar component.

georgeciubotaru commented:
Log: checking the PR

Isaacprogi commented:
hello @markholdex

markholdex reviewed (changes requested):
@Isaacprogi I noticed that the links are still rendered the `new way` in the Vercel Preview. That's because you are still parsing and replacing them within the current pages. 

---

The parsing logic should become part of the `google-conversion` API instead. So in the Preview they won't be displayed the `new way` now. Once we merge this PR on production, we can update any article and the `google-conversion` will catch the new link blocks and save them on back-end with the new style. Otherwise, having it to render in live is too resource-heavy and unreliable. 

As I said, check the example from #570 and from the `google-conversion` API. You will find many other components there.

Isaacprogi commented:
> @Isaacprogi I noticed that the links are still rendered the `new way` in the Vercel Preview. That's because you are still parsing and replacing them within the current pages.
> 
> The parsing logic should become part of the `google-conversion` API instead. So in the Preview they won't be displayed the `new way` now. Once we merge this PR on production, we can update any article and the `google-conversion` will catch the new link blocks and save them on back-end with the new style. Otherwise, having it to render in live is too resource-heavy and unreliable.
> 
> As I said, check the example from #570 and from the `google-conversion` API. You will find many other components there.
Okay

Isaacprogi commented:
> @Isaacprogi I noticed that the links are still rendered the `new way` in the Vercel Preview. That's because you are still parsing and replacing them within the current pages.
> 
> The parsing logic should become part of the `google-conversion` API instead. So in the Preview they won't be displayed the `new way` now. Once we merge this PR on production, we can update any article and the `google-conversion` will catch the new link blocks and save them on back-end with the new style. Otherwise, having it to render in live is too resource-heavy and unreliable.
> 
> As I said, check the example from #570 and from the `google-conversion` API. You will find many other components there.

Okay
